src/frontend/hooks/useLocalDiscoveryState.ts

```diff
@@ -0,0 +1,210 @@
+      return api.startLocalPeerDiscovery();
+    },
+    stop() {
+      return api.stopLocalPeerDiscovery();
+    },
+  });
+  sm.on('state', state => {
+    if (state.value === 'error') {
+      updateState({status: 'error', error: state.error});
+    } else {
+      updateState({status: state.value});
+    }
+  });
+  const listeners = new Set<() => void>();
+
+  NetInfo.addEventListener(onNetInfo);
+  AppState.addEventListener('change', onAppState);
+  refreshWifiState();
+
+  function refreshWifiState() {
+    NetInfo.fetch('wifi').then(onNetInfo).catch(noop);
+  }
+
+  function updateState(newState: Partial<LocalDiscoveryState>) {
+    if (shallowPartialEqual(state, newState)) return;
+    // For React's useSyncExternalStore, getSnapshot needs to return a different
+    // value (as compared by `Object.is`) when the state changes.
+    state = {
+      ...state,
+      ...newState,
+    };
+    listeners.forEach(listener => listener());
+  }
+
+  function onNetInfo(nextNetInfo: NetInfoState) {
+    if (state.status === 'error') return;
+    if (nextNetInfo.type !== 'wifi') {
+      // Currently NetInfo events fire with the active default data network. If
+      // the phone is connected to WiFi, but the wifi connection has no
+      // internet, and is also connected to cellular, with an active internet
+      // connection, then the callback will fire with the cellular connection,
+      // but we are interested in the wifi connection state, even if not
+      // connected to the internet.
+      refreshWifiState();
+      return;
+    }
+    // Don't do anything if the app is not active (starting and stopping
+    // discovery based on appState is handled below in onAppState)
+    if (appState !== 'active') return;
+    const nextIpAddress = nextNetInfo.details.ipAddress;
+    const prevIpAddress = netInfo?.details ? netInfo.details.ipAddress : null;
+    // netInfo.isConnected is true (I think) only if the wifi network has
+    // internet access. We use the presence of an IP address to detect if wifi
+    // is connected to a local network.
+    const isLocalWifiConnected = !!nextIpAddress;
+
+    const shouldDiscoveryBeStarted =
+      isLocalWifiConnected &&
+      (state.status === 'stopped' || state.status === 'stopping');
+
+    const ipAddressChanged =
+      nextIpAddress && prevIpAddress && nextIpAddress !== prevIpAddress;
+
+    const ssidChanged =
+      nextNetInfo.details.ssid &&
+      netInfo?.details?.ssid &&
+      nextNetInfo.details.ssid !== netInfo.details.ssid;
+
+    const shouldDiscoveryRestart =
+      (ipAddressChanged || ssidChanged) &&
+      (state.status === 'starting' || state.status === 'started');
+
+    const shouldDiscoveryBeStopped =
+      !isLocalWifiConnected &&
+      (state.status === 'starting' || state.status === 'started');
+    if (shouldDiscoveryRestart) {
+      // StartStopStateMachine will ensure these happen consecutively
+      sm.stop().catch(noop);
+      sm.start().catch(noop);
```

Comment:
what happens if you call `sm.start()` when `state === 'stopping'`? Aka, does it need to be completely stopped before we can start it again?

Reply:
When you call `sm.start()` when `state === 'stopping'`, it will wait for `stopped` before starting. The [state machine docs](https://github.com/digidem/start-stop-state-machine#api) attempt to explain the logic. Dealing this this stuff has so many edge cases it's a pain to deal with, but hopefully the start-stop-state-machine allows the code here to be a little simpler.